gemrb/plugins/SDLVideo/SDLSurfaceSprite2D.cpp

```diff
@@ -188,97 +167,122 @@ bool SDLSurfaceSprite2D::ConvertFormatTo(const PixelFormat& tofmt) noexcept
 	return false;
 }
 
-void SDLSurfaceSprite2D::Invalidate() const noexcept
-{
-	version = 0;
-	if (renderedSurface != surface) {
-		SDL_FreeSurface(renderedSurface);
-		renderedSurface = surface;
+void SDLSurfaceSprite2D::ensureShadedPalette() const noexcept {
+	if (!shadedPalette) {
+		shadedPalette = MakeHolder<Palette>();
 	}
-	if (format.Depth <= 8) {
-		UpdateSurfacePalette();
+}
+
+void SDLSurfaceSprite2D::shadePalette(BlitFlags renderflags, const Color* tint) const noexcept {
+	Palette::Colors buffer;
+
+	for (size_t i = 1; i < 256; ++i) {
+		buffer[i] = format.palette->operator[](i);
+
+		if (renderflags&BlitFlags::COLOR_MOD) {
+			ShaderTint(*tint, buffer[i]);
+		}
+
+		if (renderflags & BlitFlags::ALPHA_MOD) {
+			buffer[i].a = tint->a;
+		}
+
+		if (renderflags&BlitFlags::GREY) {
+			ShaderGreyscale(buffer[i]);
+		} else if (renderflags&BlitFlags::SEPIA) {
+			ShaderSepia(buffer[i]);
+		}
 	}
+
+	shadedPalette->CopyColors(1, buffer.cbegin() + 1, buffer.cend());
+}
+
+bool SDLSurfaceSprite2D::needToUpdatePalette() const noexcept {
+	return format.palette && palVersion != format.palette->GetVersion();
 }
 
-void* SDLSurfaceSprite2D::NewVersion(version_t newversion) const noexcept
+bool SDLSurfaceSprite2D::needToReshade(const Color* tint, BlitFlags flags) const noexcept {
+	return flags != appliedBlitFlags
+		|| ((flags & BlitFlags::COLOR_MOD) && tint && (appliedTint.Packed() & 0xFFFFFF00) != (tint->Packed() & 0xFFFFFF00))
+		|| ((flags & BlitFlags::ALPHA_MOD) && tint && (appliedTint.Packed() & 0xFF) != (tint->Packed() & 0xFF))
+		|| needToUpdatePalette()
+		|| surfaceInvalidated;
```

Comment:
This could be checked sooner, for a tiny speed increase.

Reply:
Sure, why not